### IMPORTS

In [1]:
import numpy as np
import scipy as scp
import itertools

$\textbf{QUESTION 1}$

$$
\textbf{PART A}
\\~\\
A = \left[ 0, \  0\right]\\
B = \left[2, \ 0\right]\\
C = \left[1, \ 2\right]\\
x = \left[1, \ \frac{2}{3}\right]
\\~\\
x = \lambda_{1} * A +  \lambda_{2} * B + \lambda_{3} *  C\\
\lambda_{1} +  \lambda_{2} + \lambda_{3} = 1
\\~\\
\text{We know that since C has the only non zero value in the second row (if transposed), then}\ \lambda{3} * 2 = \frac{2}{3}\\ \Rightarrow \lambda{1} = \frac{1}{3}\\~\\ \Rightarrow \lambda{2} = \frac{1}{3}\\~\\\Rightarrow \lambda{3} = \frac{1}{3}\\
$$

$$
\textbf{PART B}
\\~\\
\text{Extreme Points:}\\
x_1 = [1, 1]\\
x_2 = [5, 1]
\\~\\
\text{Extreme Rays:}\\
e_1 = [-1, 1]\\
e_2 = [0, 1]
$$

<p align="center">
<img src="/home/leonardo_leads/Documents/SchoolDocs/ga_tech_masters/omscs_ml/fall2023/deterministic_optimization/lectures/week7Modules13and14/homework/images/q1b.png" width="600" height="600" />
</p>

$$
\textbf{PART C}
\\~\\
\text{The maximum number of extreme points is 4. There are several planes that could be used but one such plane has the following extreme points:}
\\~\\
x_1 = [0, 0, 0]\\
x_2 = [1, 0, 0]\\
x_3 = [0, 1, 1]\\
x_4 = [1, 1, 1]
\\~\\
\text{From the below graph you could imagine the plane passing through the points:}
$$
<p align="center">
<img src="/home/leonardo_leads/Documents/SchoolDocs/ga_tech_masters/omscs_ml/fall2023/deterministic_optimization/lectures/week7Modules13and14/homework/images/q1c.png" width="600" height="600" />
</p>


$\textbf{QUESTION 2}$

$$
\textbf{PART A}
\\~\\
\textbf{Standard Form Linear Program:}\\
\textbf{Objective Function:}\\
\text{-min}\ x+y-z
\\~\\
\textbf{Constraints:}\\
-x+y-z-s_1 = 1\\
-2x-y+z+s_2 = 1\\
x,y,z,s_1,s_2 \ge 0\\
$$

$$
\textbf{PART B}
\\~\\
\textbf{Linear Program Formulation:}\\
\textbf{Objective Function:}\\
\text{min}\ z_1 + z_2
\\~\\
\textbf{Constraints:}\\
x+y \le z_1 \\
x-2y \le z_2 \\
x \le 10 \\
y \le 10 \\
x+y \le 10 \\
z_1 \ge 0 \\
z_2 \ge 0\\
x+y \ge -z_1\\
x-2y \ge -z_2
$$

$\textbf{QUESTION 3}$
$$
\textbf{PART A}\\
\text{The feasible region is the below shaded region with extreme points:}\\
p_1 = [0, 0]\\
p_2 = [3, 0]\\
p_3 = [2, 2]\\
p_4 = [0.667, 3.333]\\
p_5 = [0, 2]
$$
<p align="center">
<img src="/home/leonardo_leads/Documents/SchoolDocs/ga_tech_masters/omscs_ml/fall2023/deterministic_optimization/lectures/week7Modules13and14/homework/images/q3c.png" width="600" height="600" />
</p>

$$
\\~\\
\textbf{PART B}\\
A =
\begin{bmatrix}
1 & 1 & 1 & 0 & 0\\
-2 & 1 & 0 & 1 & 0\\
2&1&0&0&1
\end{bmatrix}
\\~\\
c = 
\begin{bmatrix}
-1 & -2 & 0 & 0 & 0
\end{bmatrix}
\\~\\
b = 
\begin{bmatrix}
4 & 2 & 6
\end{bmatrix}
$$

$$
\textbf{PART C}
$$

In [10]:
# Create A Matrix
A = [[1, 1, 1, 0, 0],
     [-2, 1, 0, 1, 0],
     [2, 1, 0, 0, 1]]
A = np.matrix(A)

# Create B Vector
b = [4, 2, 6]
b = np.array(b)

# Non Basic Variables
nonbasic_var = [0, 0]

# Basic and Nonbasic Cost Coefficients
cost_coeff = [-1, -2, 0, 0, 0]
cost_coeff = np.array(c)
basic_cost_coeff = c[:2]
nonbasic_cost_coeff = c[2:]

for i, j, k in itertools.combinations([0,1,2,3,4], 3):
    basis = A[:, [i, j, k]]
    basic_var = scp.linalg.inv(basis)@b
    print(f"\nB =\n{basis}\nx_b = {basic_var}\nx_n = {nonbasic_var}\nc_b = {basic_cost_coeff}\nc_n = {nonbasic_cost_coeff}")


B =
[[ 1  1  1]
 [-2  1  0]
 [ 2  1  0]]
x_b = [ 1.  4. -1.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[ 1  1  0]
 [-2  1  1]
 [ 2  1  0]]
x_b = [2. 2. 4.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[ 1  1  0]
 [-2  1  0]
 [ 2  1  1]]
x_b = [0.66666667 3.33333333 1.33333333]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[ 1  1  0]
 [-2  0  1]
 [ 2  0  0]]
x_b = [3. 1. 8.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[ 1  1  0]
 [-2  0  0]
 [ 2  0  1]]
x_b = [-1.  5.  8.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[ 1  0  0]
 [-2  1  0]
 [ 2  0  1]]
x_b = [ 4. 10. -2.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[1 1 0]
 [1 0 1]
 [1 0 0]]
x_b = [ 6. -2. -4.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[1 1 0]
 [1 0 0]
 [1 0 1]]
x_b = [2. 2. 4.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[1 0 0]
 [1 1 0]
 [1 0 1]]
x_b = [ 4. -2.  2.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 0]

B =
[[1 0 0]
 [0 1 0]
 [0 0 1]]
x_b = [4. 2. 6.]
x_n = [0, 0]
c_b = [-1 -2]
c_n = [0 0 

$$
\textbf{PART D}
\\~\\
\text{Feasible Solutions:}\\
x =
\begin{bmatrix}
2 & 2 & 0 & 4 & 0
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
0.667 & 3.33 & 0 & 0 & 1.33
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
3 & 0 & 1 & 8 & 0
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
0 & 2 & 2 & 0 & 4
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
0 & 0 & 4 & 2 & 6
\end{bmatrix}

\\~\\
\text{Infeasible Solutions:}\\
\\~\\
x =
\begin{bmatrix}
1 & 4 & -1 & 0 & 0
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
-1 & 0 & 5 & 0 & 8
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
4& 0 & 0 & 10 & -2
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
0 & 6 & -2 & -4 & 0
\end{bmatrix}
\\~\\
x =
\begin{bmatrix}
0 & 4 & 0 & -2 & 2
\end{bmatrix}
\\~\\
\text{The Basic Feasible Solutions are all of the Extreme Points in the graph in Part A.}\\
\text{The infeasible solutions are all intersections in the above graph, but are outside of the feasible (shaded) region above.}
$$